File: BeakJoonT/토마토.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

vector<vector<int>> v;
int M = 0, N = 0;
int move_h[4] = { 0, 0, 1, -1 };
int move_w[4] = { -1, 1, 0, 0 };
bool visit[1001][1001] = { false };
queue<pair<pair<int, int>, int>> q;
int answer = 0;

void bfs() {
	while (!q.empty()) {
		int h = q.front().first.first;
		int w = q.front().first.second;
		int time = q.front().second;
		answer = time;
		q.pop();

		for (int i = 0; i < 4; i++) {
			int tmp_w = w + move_w[i];
			int tmp_h = h + move_h[i];
			if (tmp_w >= 0 && tmp_h >= 0 && tmp_w < M && tmp_h < N) {
				if (v[tmp_h][tmp_w] == 0 && !visit[tmp_h][tmp_w]) {
					v[tmp_h][tmp_w] = 1;
					visit[tmp_h][tmp_w] = true;
					q.push(make_pair(make_pair(tmp_h, tmp_w), time + 1));
				}
			}
		}

	}
}

int main()
{	
	ios_base::sync_with_stdio(false), cin.tie(nullptr), cout.tie(nullptr);
	cin >> M >> N;

	for (int i = 0; i < N; i++) {
		vector<int> tmp;
		for (int j = 0; j < M; j++) {
			int n = 0;
			cin >> n;
			tmp.push_back(n);
		}
		v.push_back(tmp);
	}

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (!visit[i][j] && v[i][j] == 1) {
				visit[i][j] = true;
				q.push(make_pair(make_pair(i, j), 0));
			}
		}
	}

	bfs();

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			if (v[i][j] == 0) answer = -1;
		}
	}

	cout << answer;
}

File: BeakJoonT/토마토.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Q7569 {
    static int N, M, H;
    static int[][][] board;
    static int[][][] cnt;
    static int[] moveLayer = {0, 0, 0, 0, -1, 1};
    static int[] moveH = {-1, 1, 0, 0, 0, 0};
    static int[] moveW = {0, 0, 1, -1, 0, 0};
    static Queue<int []> que = new LinkedList<>();

    static void bfs(){
        while (!que.isEmpty()) {
            int[] tmp = que.poll();

            for (int i = 0; i < 6; i++){
                int tmpLayer = tmp[0] + moveLayer[i];
                int tmpH = tmp[1] + moveH[i];
                int tmpW = tmp[2] + moveW[i];

                if(tmpH < 0 || tmpW < 0 || tmpH >= N || tmpW >= M || tmpLayer < 0 || tmpLayer >= H
                        || board[tmpLayer][tmpH][tmpW] != 0) continue;

                board[tmpLayer][tmpH][tmpW] = 1;
                cnt[tmpLayer][tmpH][tmpW] = cnt[tmp[0]][tmp[1]][tmp[2]] + 1;

                que.offer(new int[] {tmpLayer, tmpH, tmpW});
            }
        }

        int answer = Integer.MIN_VALUE;

        for (int h = 0; h < H; h++){
            for (int i = 0; i < N; i++){
                for (int j = 0; j < M; j++){
                    if(board[h][i][j] == 0) {
                        System.out.println(-1);
                        return;
                    }
                    answer = Math.max(answer, cnt[h][i][j]);
                }
            }
        }

        if(answer == 1) System.out.println(0);
        else System.out.println(answer - 1);
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());

        board = new int[H][N][M];
        cnt = new int[H][N][M];

        for (int h = 0; h < H; h++){
            for (int i = 0; i < N; i++){
                st = new StringTokenizer(br.readLine());
                for (int j = 0; j < M; j++){
                    board[h][i][j] = Integer.parseInt(st.nextToken());
                    if(board[h][i][j] == 1) cnt[h][i][j] = 1;
                }
            }
        }

        for (int h = 0; h < H; h++){
            for (int i = 0; i < N; i++){
                for (int j = 0; j < M; j++){
                    if(board[h][i][j] == 1) que.offer(new int[] {h, i, j});
                }
            }
        }

        bfs();
    }
}
